modules/validations/outdated_tags.js

```diff
@@ -93,63 +160,120 @@ export function validationOutdatedTags() {
-      }
 
-      // try key/value|name match against name-suggestion-index
-      if (newTags.name) {
-        for (let i = 0; i < nsiKeys.length; i++) {
-          const k = nsiKeys[i];
-          if (!newTags[k]) continue;
-
-          const center = entity.extent(graph).center();
-          const countryCode = countryCoder.iso1A2Code(center);
-          const match = _nsi.matcher.matchKVN(k, newTags[k], newTags.name, countryCode && countryCode.toLowerCase());
-          if (!match) continue;
-
-          // for now skip ambiguous matches (like Target~(USA) vs Target~(Australia))
-          if (match.d) continue;
-
-          const brand = _nsi.brands[match.kvnd];
-          if (brand && brand.tags['brand:wikidata'] &&
-            brand.tags['brand:wikidata'] !== entity.tags['not:brand:wikidata']) {
-            subtype = 'noncanonical_brand';
-
-            const keepTags = ['takeaway'].reduce((acc, k) => {
-              if (newTags[k]) {
-                acc[k] = newTags[k];
-              }
+        if (!loc) {     // collect location for this feature only once
+          loc = entity.extent(graph).center();
+        }
+        if (!names) {   // collect names for this feature only once
+          names = Object.keys(newTags)
+            .map(k => isNamelike(k) ? newTags[k] : null)
+            .filter(Boolean);
+
+          if (foundQID) names.unshift(foundQID);  // matcher will recognize the QID as a name too
+          names = utilArrayUniq(names);
+        }
+
+        // Try each namelike value
+        for (let n = 0; n < names.length; n++) {
+          match = _nsi.matcher.match(k, v, names[n], loc);   // Attempt to match an item in NSI
+          if (!match) continue;  // keep looking
+
+          // If we get here, there was a match..
+          // A match may contain multiple results, the first one is the best one for this location
+          // e.g. `['pfk-a54c14', 'kfc-1ff19c', 'kfc-658eea']`
+          const itemID = match[0].itemID;
+          const item = _nsi.ids.get(itemID);
+          const mainTag = item.mainTag;               // e.g. `brand:wikidata`
+          const itemQID = item.tags[mainTag];         // e.g. `brand:wikidata` qid
+          const notQID = newTags[`not:${mainTag}`];   // e.g. `not:brand:wikidata` qid
+
+          // Exceptions, throw out the match
+          if (
+            (!itemQID || itemQID === notQID) ||       // no `*:wikidata` or matched a `not:*:wikidata`
+            (newTags.office && !item.tags.office)     // feature may be a coprorate office for a brand? - #6416
+          ) {
+            match = null;   // forget match and keep looking
+            continue;       // (it might make sense to stop looking, not sure)
+          }
+
+          // We are keeping the match at this point
+          subtype = 'noncanonical_brand';
+
+          // Preserve some tags values that we don't want NSI to overwrite.
+          const keepTags = ['takeaway', 'building']
```

Comment:
@1ec5 we should add `flag:name` here to leave people's local flag names alone.

Reply:
I did this in f61a3ef5e, and it works:

before | after
--- | ---
<img width="350" alt="Screen Shot 2021-01-18 at 9 37 04 PM" src="https://user-images.githubusercontent.com/38784/104981433-9a357480-59d6-11eb-998a-5c4b8126f355.png"> | <img width="350" alt="Screen Shot 2021-01-18 at 9 42 27 PM" src="https://user-images.githubusercontent.com/38784/104981447-a0c3ec00-59d6-11eb-859a-d76eb0403003.png">